### ISSUE TITLE
Friendly URLs in DNN must be on

### ISSUE THREAD

**Issue by leedavi** [openstore-ecommerce/OpenStore#105]
It's been reported that friendly URLs must be on for the links to the detail page to work.

This should work in both friendly and non-friendly url mode.

**Comment by trouble2** [comment]
Well, what I sometimes see is the link on the list page is showing a friendly url while the detail page isn't like here:
https://www.vankortenhof.nl/fietsen
when I hover over the products I can see links like :
https://www.vankortenhof.nl/fietsen/p/catid/56/eid/15318/van-kortenhof-liv-allure-cs2
However when I click on them it changes to:
https://www.vankortenhof.nl/fietsen/p/catid/56/eid/15318

Any idea about this? Looks more like a DNN issue...

**Comment by DNNMonster** [comment]
I am pretty sure that the humanfriendly and searchfriendly options for the DNNFriendlyUrl provider require the .aspx extension in the url to properly render.

Since the urls generated by OS are not generating urls with the .aspx extension it results in the files getting processed by the StaticFileHandler and thus http 404 errors.

To verify you can change your web config to use the humanfriendly or searchfriendly options. Navigate to a detail page.  Once you land on a detail page it should throw a 404.  If you add a .aspx to the end of that failed page then it will render.

If those config options are to be supported my first thought is that OS needs to detect this and add the .aspx extension when generating urls.

**Comment by trouble2** [comment]
One is DNN 9.8.0, the other 9.8.1
They both use: urlFormat="advanced"
And one works and the other one doesn't...

**Comment by DNNMonster** [comment]
If I understand things correctly I think 2 issues a being discussed here.

1. OS needs to add .aspx to it's urls to support the humanfriendly and searchfriendly DNNFriendlyUrl options. (That's a possible solution for this thread)

2. A dev needs to debug the NBrightBuyOpenUrlRewriter project for a better understanding of why installations that are using it are having problems with generating a consistent details page url pattern that matches the list view.  This could have it's own ticket opened up under the NBrightBuyOpenUrlRewriter project to ensure it doesn't slip through the cracks.

**Comment by leedavi** [comment]
I'm no expert on DNN URL formats, but from what I understand "NBrightBuyOpenUrlRewriter" is not being used.

However, many years ago I did write the function responsible for generating the URL.

        public static string GetEntryUrl(int portalId, string entryid, string modulekey, string seoname, string tabid, string catid = "", string catref = "")


This builds a standard URL, and then uses the DNN function to generate a friendly version.

            urldata = DotNetNuke.Services.Url.FriendlyUrl.FriendlyUrlProvider.Instance().FriendlyUrl(objTabInfo, strurl, seoname);


There is some caching on this function, so maybe the old url is cached?  A restart of the app Pool should prove this idea correct or false.

But, it does look like DNN is not generating the correct friendly URL.  I've never had an issue with the URL format, but  it looks like some setting in DNN has been changed?

We have a policy of changing as little as possible in a standard config.  So we have this in the web.config.

    <friendlyUrl defaultProvider="DNNFriendlyUrl">
      <providers>
        <clear />
        <add name="DNNFriendlyUrl" type="DotNetNuke.Services.Url.FriendlyUrl.DNNFriendlyUrlProvider, DotNetNuke.HttpModules" includePageName="true" regexMatch="[^a-zA-Z0-9 _-]" urlFormat="advanced" />
      </providers>
    </friendlyUrl>

The URL Friendly functionality in DNN does use the portal Alias information from the portal  We find this can often create a problem.  Check your portal alias is correct and try and use only 1 if possible.

Also, because of performance we never turn on forced SSL in DNN, we always use the web.config to do a redirect from http to https.
